once/__init__.py

```diff
@@ -24,30 +26,57 @@ class _OnceBase(abc.ABC):
     """Abstract Base Class for once function decorators."""
 
     def __init__(self, func: collections.abc.Callable):
-        self._inspect_function(func)
         functools.update_wrapper(self, func)
-        self.lock = _new_lock()
+        self.func = self._inspect_function(func)
         self.called = False
         self.return_value: typing.Any = None
-        self.func = func
+        self.is_asyncgen = inspect.isasyncgenfunction(self.func)
+        self.is_async = True if self.is_asyncgen else inspect.iscoroutinefunction(self.func)
+        if self.is_async:
+            self.async_lock = asyncio.Lock()
+        else:
+            self.lock = _new_lock()
 
     @abc.abstractmethod
-    def _inspect_function(self, func: collections.abc.Callable):
+    def _inspect_function(self, func: collections.abc.Callable) -> collections.abc.Callable:
         """Inspect the passed-in function to ensure it can be wrapped.
 
         This function should raise a SyntaxError if the passed-in function is
-        not suitable."""
+        not suitable.
 
-    def _execute_call_once(self, func: collections.abc.Callable, *args, **kwargs):
+        It should return the function which should be executed once.
+        """
+
+    async def _execute_call_once_async(self, func: collections.abc.Callable, *args, **kwargs):
+        if self.called:
+            return self.return_value
+        async with self.async_lock:
+            if self.called:
+                return self.return_value
+            if self.is_asyncgen:
+                self.return_value = tuple([i async for i in func(*args, **kwargs)])
```

Comment:
This is a big caveat that needs detailing in the documentation, that the generator result gets converted to a tuple. I think the more correct version, that's also more work, would wrap the generator in a generator that extends the return_value list each time next is called. This is potentially surprising enough to call it a bug that warrants exclusion from this library until it's fully supported.